test/v1/services/SelfEmploymentPropertyBISSServiceSpec.scala

```diff
@@ -0,0 +1,105 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1.services
+
+import support.UnitSpec
+import uk.gov.hmrc.domain.Nino
+import uk.gov.hmrc.http.HeaderCarrier
+import v1.controllers.EndpointLogContext
+import v1.mocks.connectors.MockSelfEmploymentPropertyBISSConnector
+import v1.models.des.IncomeSourceType
+import v1.models.errors._
+import v1.models.outcomes.ResponseWrapper
+import v1.models.requestData.{DesTaxYear, RetrieveUKPropertyBISSRequest}
+import v1.models.response.RetrieveUKPropertyBISSResponse
+import v1.models.response.common.{Loss, Profit, Total}
+
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+class SelfEmploymentPropertyBISSServiceSpec extends UnitSpec {
+
+  private val nino = "AA123456A"
+  private val taxYear = "2019"
+  private val selfEmploymentId = "123456789"
+  private val correlationId = "X-123"
+
+  private val requestData = RetrieveUKPropertyBISSRequest(Nino(nino), DesTaxYear(taxYear), IncomeSourceType.`uk-property`)
+
+  trait Test extends MockSelfEmploymentPropertyBISSConnector {
+    implicit val hc: HeaderCarrier = HeaderCarrier()
+    implicit val logContext: EndpointLogContext = EndpointLogContext("controller", "retrieveSelfEmploymentPropertyBISS")
+
+    val service = new SelfEmploymentPropertyBISSService(mockConnector)
+  }
+
+  val response =
+    RetrieveUKPropertyBISSResponse(
+      Total(
+        income = 100.00,
+        expenses = Some(50.00),
+        additions = Some(5.00),
+        deductions = Some(60.00)
+      ),
+      accountingAdjustments = Some(-30.00),
+      Some(Profit(
+        net = Some(20.00),
+        taxable = Some(10.00)
+      )),
+      Some(Loss(
+        net = Some(10.00),
+        taxable = Some(35.00)
+      ))
+    )
+
+  "retrieveBiss" should {
+    "return a valid response" when {
+      "a valid request is supplied" in new Test {
+        MockSelfEmploymentBISSPropertyConnector.retrieveBiss(requestData)
+          .returns(Future.successful(Right(ResponseWrapper(correlationId, response))))
+
+        await(service.retrieveBiss(requestData)) shouldBe Right(ResponseWrapper(correlationId, response))
+      }
+    }
+
+    "return error response" when {
+
+      def serviceError(desErrorCode: String, error: MtdError): Unit =
+        s"a $desErrorCode error is returned from the service" in new Test {
+
+          MockSelfEmploymentBISSPropertyConnector.retrieveBiss(requestData)
+            .returns(Future.successful(Left(ResponseWrapper(correlationId, DesErrors.single(DesErrorCode(desErrorCode))))))
+
+          await(service.retrieveBiss(requestData)) shouldBe Left(ErrorWrapper(Some(correlationId), error))
+        }
+
+      val input = Seq(
+
+        ("INVALID_IDVALUE", NinoFormatError),
+        ("INVALID_INCOMESOURCEID", DownstreamError),
```

Comment:
Missing  "INVALID_TAXYEAR" -> TaxYearFormatError scenario... It's easy if we start with 400's, then 403 (if any), 404 and lastly 500's